### PR TITLE
Pretty Print when using rules (#440)

### LINKED ISSUE #440: Rule is not supported
### What version of godog are you using?
```
> godog version: 0.12.2
```

### What version of Go are you using?
```
> go version: 1.16
```

### What did you do?
Given a feature file like the following:

```
Feature: simple feature
  simple feature description

  Rule: simple rule
    Example: simple example
```

Running `godog` will panic:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x70 pc=0x11cc0c6]

goroutine 34 [running]:
github.com/cucumber/godog/internal/formatters.(*Pretty).printStep(0xc0002020d0, 0xc000250360, 0xc000208680)
        /Users/michaelsauter/go/pkg/mod/github.com/cucumber/godog@v0.12.2/internal/formatters/fmt_pretty.go:384 +0xf26
``` 

### Additional context
I've looked a bit into the source code and added the following test to `pretty.feature`:
```
  Scenario: Support of Feature Plus Rule Node
    Given a feature "features/simple.feature" file:
    """
        Feature: simple feature
            simple feature description

        Rule: simple rule
            Scenario: simple scenario
                simple scenario description
    """
    When I run feature suite with formatter "pretty"
    Then the rendered output will be as follows:
    """
      Feature: simple feature
        simple feature description

        Rule: simple rule # features/simple.feature:3

        Scenario: simple scenario # features/simple.feature:4

      1 scenarios (1 undefined)
      No steps
      0s
    """
```

This lead me to apply the following fix in `fmt_pretty.go`:
```go
func (f *Pretty) scenarioLengths(pickle *messages.Pickle) (scenarioHeaderLength int, maxLength int) {
	feature := f.Storage.MustGetFeature(pickle.Uri)
	astRule := feature.FindRule(pickle.AstNodeIds[0])
	astScenario := feature.FindScenario(pickle.AstNodeIds[0])
	astBackground := feature.FindBackground(pickle.AstNodeIds[0])

	if astScenario != nil {
		scenarioHeaderLength = f.lengthPickle(astScenario.Keyword, astScenario.Name)
		maxLength = f.longestStep(astScenario.Steps, scenarioHeaderLength)
	}

	if astBackground != nil {
		maxLength = f.longestStep(astBackground.Steps, maxLength)
	}
	if astRule != nil {
		for _, rc := range astRule.Children {
			if rc.Scenario != nil {
				maxLength = f.longestStep(rc.Scenario.Steps, maxLength)
			} else if rc.Background != nil {
				maxLength = f.longestStep(rc.Scenario.Steps, maxLength)
			}
		}
	}

	return scenarioHeaderLength, maxLength
}
```

... but there are many more issues in the pretty formatter, plus I realized all the other formatters would need to be fixed, too. Seems like supporting rules will be quite the effort. Are there plans to do this already?

### PR DESCRIPTION
**Pull request by dumpsterfireproject**
### 🤔 What's changed?

- updated internal/models/feature.go to also find scenarios, backgrounds, and steps that occur under rules
- updated internal/formatters/fmt_pretty.go to also output backgrounds that occur under rules
- updated feature/formatter/pretty.feature to include some simple examples with rules
- updated run_test.go to account for new scenarios added in this PR
- added internal/formatters/formatter-tests/features/rules_with_examples_with_backgrounds.feature and internal/formatters/formatter-tests/pretty/rules_with_examples_with_backgrounds to provide additional testing of this change

### ⚡️ What's your motivation? 

Fixes #440 

### 🏷️ What kind of change is this?

- :bug: Bug fix (non-breaking change which fixes a defect)

### ♻️ Anything particular you want feedback on?

The description in #440 focused on the pretty formatter. I kept the scope for this PR limited to the pretty formatter. I can add tests and look into any issues that come up with rules with the other formatters. My thought was to limit scope for this PR to just the pretty formatter, but if you want a different approach, just let me know. If there are additional tests that are needed, please let me know that as well.

### 📋 Checklist:

- [x] I agree to respect and uphold the [Cucumber Community Code of Conduct](https://cucumber.io/conduct/)
- [x] I've changed the behaviour of the code
  - [x] I have added/updated tests to cover my changes.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly.
- [ ] Users should know about my change
  - [ ] I have added an entry to the "Unreleased" section of the [**CHANGELOG**](../CHANGELOG.md), linking to this pull request.

### CODE REVIEW

File: features/formatter/pretty.feature [hunk 1]
```diff
@@ -449,3 +449,58 @@ Feature: pretty formatter
     16 steps (16 passed)
     0s
     """
+
+  Scenario: Support of Feature Plus Rule
+      Given a feature "features/simple.feature" file:
```
Comment (mattwynne) [inline features/formatter/pretty.feature:454]: Indentation of this `Given` looks a bit off?
```suggestion
    Given a feature "features/simple.feature" file:
```

File: features/formatter/pretty.feature [hunk 2]
```diff
@@ -449,3 +449,58 @@ Feature: pretty formatter
...
+
+        Example: simple scenario # features/simple.feature:5
+          Given passing step     # suite_context.go:0 -> SuiteContext.func2
+
+      1 scenarios (1 passed)
+      1 steps (1 passed)
+      0s
+    """
+
+Scenario: Support of Feature Plus Rule with Background
```
Comment (mattwynne) [inline features/formatter/pretty.feature:478]:
```suggestion
  Scenario: Support of Feature Plus Rule with Background
```

Comment (dumpsterfireproject) [reply features/formatter/pretty.feature]: Just pushed that change.  Thanks.

**Comment by dumpsterfireproject** [comment]
I added a test for a rule with a scenario outline as well.

**Comment by mattwynne** [comment]
Looks good to me! Any feedback @vearutop?

**Review by mattwynne — APPROVED** [review]

**Comment by aslakhellesoy** [comment]
Hi @dumpsterfireproject,

Thanks for your making your first contribution to Cucumber, and welcome to the Cucumber committers team! You can now push directly to this repo and all other repos under the cucumber organization! 🍾

In return for this generous offer we hope you will:

* ✅ Continue to use branches and pull requests. When someone on the core team approves a pull request (yours or someone else's), you're welcome to merge it yourself.
* 💚 Commit to setting a good example by following and upholding our [code of conduct](https://cucumber.io/conduct) in your interactions with other collaborators and users.
* 💬 Join the [community Slack channel](http://cucumber.io/support#slack) to meet the rest of the team and make yourself at home.
* ℹ️ Don't feel obliged to help, just do what you can if you have the time and the energy.
* 🙋 Ask if you need anything. We're looking for feedback about how to make the project more welcoming, so please tell us!

On behalf of the Cucumber core team,
Aslak Hellesøy
Creator of Cucumber

**Comment by mattwynne** [comment]
Welcome to the team @dumpsterfireproject!

You're now able to push directly to branches within the https://github.com/cucumber/godog repo, so no need to use your own fork now unless you particularly want to.

This also means you have the permissions to review and merge other people's pull requests. If you have any time or inclination for that, it would be greatly appreciated!

**Comment by mattwynne** [comment]
Oops, we forgot to update the `CHANGELOG.md` - can you do that, @dumpsterfireproject? You could make a commit direct to `main` for something like that, rather than making another PR.

**Comment by dumpsterfireproject** [comment]
Sure, I'll get the CHANGELOG.md updated.  Thanks.

**Pull request merged**